src/coreComponents/common/LifoStorage.hpp

```diff
@@ -0,0 +1,198 @@
+#include <unistd.h>
+#include <algorithm>
+
+#include "common/GEOS_RAJA_Interface.hpp"
+#include "common/TimingMacros.hpp"
+#include "common/LifoStorageCommon.hpp"
+#include "common/LifoStorageHost.hpp"
+#ifdef GEOSX_USE_CUDA
+#include "common/LifoStorageCuda.hpp"
+#endif
+
+namespace geosx
+{
+/**
+ * This class is used to store in a LIFO way buffers, first on device, then on host, then on disk.
+ */
+template< typename T, typename INDEX_TYPE >
+class LifoStorage
+{
+
+private:
+  /// number of buffers to be inserted into the LIFO
+  int m_maxNumberOfBuffers;
+  /// size of one buffer in bytes
+  size_t m_bufferSize;
+  /// counter of buffer stored in LIFO
+  int m_bufferCount;
+
+  /// pointer either to CUDA aware or host only LifoStorage
+  std::unique_ptr< LifoStorageCommon< T, INDEX_TYPE > > m_lifo;
+public:
+
+
+  /**
+   * A LIFO storage will store numberOfBuffersToStoreDevice buffer on
+   * deevice, numberOfBuffersToStoreHost on host and the rest on disk.
+   *
+   * @param name                           Prefix of the files used to save the occurenncy of the saved buffer on disk.
+   * @param elemCnt                        Number of elments in the LvArray we want to store in the LIFO storage.
+   * @param numberOfBuffersToStoreOnDevice Maximum number of array to store on device memory ( -1 = use 80% of remaining memory ).
+   * @param numberOfBuffersToStoreOnHost   Maximum number of array to store on host memory ( -1 = use 80% of remaining memory ).
+   * @param maxNumberOfBuffers             Number of arrays expected to be stores in the LIFO.
+   */
+  LifoStorage( std::string name, size_t elemCnt, int numberOfBuffersToStoreOnDevice, int numberOfBuffersToStoreOnHost, int maxNumberOfBuffers ):
+    m_maxNumberOfBuffers( maxNumberOfBuffers ),
+    m_bufferSize( elemCnt*sizeof( T ) ),
+    m_bufferCount( 0 )
+  {
+    LIFO_LOG_RANK( " LIFO : maximum size "<< m_maxNumberOfBuffers << " buffers " );
+    double bufferSize = ( ( double ) m_bufferSize ) / ( 1024.0 * 1024.0 );
+    LIFO_LOG_RANK( " LIFO : buffer size "<< bufferSize << "MB" );
+#ifndef GEOSX_USE_CUDA
+    numberOfBuffersToStoreOnDevice = 0;
+#else
+    if( numberOfBuffersToStoreOnDevice == -1 )
+    {
+      size_t free, total;
+      GEOSX_ERROR_IF( cudaSuccess != cudaMemGetInfo( &free, &total ), "Error getting CUDA device available memory" );
+      double freeGB = ( ( double ) free ) / ( 1024.0 * 1024.0 * 1024.0 );
+      LIFO_LOG_RANK( " LIFO : available memory on device " << freeGB << " GB" );
+      numberOfBuffersToStoreOnDevice = std::min( (int)( 0.8 * free / m_bufferSize ), m_maxNumberOfBuffers );
+    }
+#endif
+    if( numberOfBuffersToStoreOnHost == -1 )
+    {
+      size_t free = sysconf( _SC_AVPHYS_PAGES ) * sysconf( _SC_PAGESIZE );
+      numberOfBuffersToStoreOnHost = std::max( 1, std::min( (int)( 0.8 * free / m_bufferSize ), m_maxNumberOfBuffers - numberOfBuffersToStoreOnDevice ) );
+      double freeGB = ( ( double ) free ) / ( 1024.0 * 1024.0 * 1024.0 );
```

Comment:
Formula does not take into account the multiple MPI per node that share the memory, need to update that.

Reply:
I'll improve that by dividing by the size of the node communicator but it does not take care of the case where several jobs are on the same node. Is there a geosx parameter to set an arbitrary limit on the available memory ? In that case I could get this capped memory per rank, remove process RSS and get the percentage.

Reply:
I'll have the same issue if I oversubscribe on the GPU.

Reply:
I didn't copied your code correctly it seems (and running on my home directory didn't help...) No it seems OK, I removed the std::make_tuple(std::ref(mutexes)) and used your make_multilock(), tests passed fine.